compiler/rustc_serialize/src/leb128.rs

```diff
@@ -53,21 +53,29 @@ impl_write_unsigned_leb128!(write_usize_leb128, usize);
 macro_rules! impl_read_unsigned_leb128 {
     ($fn_name:ident, $int_ty:ty) => {
         #[inline]
-        pub fn $fn_name(slice: &[u8]) -> ($int_ty, usize) {
-            let mut result = 0;
-            let mut shift = 0;
-            let mut position = 0;
-            loop {
-                let byte = slice[position];
-                position += 1;
-                if (byte & 0x80) == 0 {
-                    result |= (byte as $int_ty) << shift;
-                    return (result, position);
-                } else {
+        pub fn $fn_name(slice: &[u8], position: &mut usize) -> $int_ty {
+            #[cold]
+            #[inline(never)]
+            fn do_the_rest(slice: &[u8], position: &mut usize, byte: u8) -> $int_ty {
+                let mut result = (byte & 0x7F) as $int_ty;
+                let mut shift = 7;
+                loop {
+                    let byte = slice[*position];
+                    *position += 1;
                     result |= ((byte & 0x7F) as $int_ty) << shift;
+                    if (byte & 0x80) == 0 {
+                        return result;
+                    }
+                    shift += 7;
                 }
-                shift += 7;
             }
+
+            // The first iteration of the loop is unpeeled because this code is
+            // hot and integers less than 128 are very common, occurring 80%+
+            // of the time.
```

Comment:
Is that true for all types? IIRC, that's very common for `usize` but much less so for `u64`, for example. Maybe it would make sense to have two implementations, one optimized for small values and another optimized for larger values?

Reply:
Indeed it could be a good idea to generalize the impl for various "categories" of numbers (often small, often large), it helped recently in hashing: https://github.com/rust-lang/rust/pull/92103.

Reply:
Good point! I did some follow-up investigation, and saw that this accounts for the regressions seen in some cases. For example, the `ctfe-stress-4 check incr-unchanged` regression seen on the CI run has this distribution of leb128 lengths for read `u64` values:
```
2361790 counts:
(  1)  1311716 (55.5%, 55.5%): u64 1
(  2)  1049356 (44.4%,100.0%): u64 10
(  3)      574 ( 0.0%,100.0%): u64 9
(  4)      115 ( 0.0%,100.0%): u64 2
(  5)       20 ( 0.0%,100.0%): u64 4
(  6)        5 ( 0.0%,100.0%): u64 3
(  7)        3 ( 0.0%,100.0%): u64 8
(  8)        1 ( 0.0%,100.0%): u64 7
```